proto/amino/amino.proto

```diff
@@ -76,4 +76,15 @@ extend google.protobuf.FieldOptions {
   // out := AminoJSONEncoder(&f);
   // out == {"baz":""}
   bool dont_omitempty = 11110005;
+
+  // oneof_type_name sets the type name for the given field oneof field.  This is used
+  // by the Amino JSON encoder to encode the type of the oneof field, and must be the same string as
+  // The RegisterConcrete() method usage used to register the concrete type.
+  string oneof_type_name = 11110006;
 }
+
+extend google.protobuf.OneofOptions {
+  // oneof_field_name sets the field name for the given oneof field.  This is used by the Amino JSON
+  // encoder to set the field name encapsulating the oneof field.
+  string oneof_field_name = 11110007;
```

Comment:
Consider StakeAuthorization:
https://github.com/cosmos/cosmos-sdk/blob/5f0db4d9d8181e851e00234de00455f227db4613/tests/integration/aminojson/aminojson_test.go#L494-L503
The legacy encoder traverses the type with golang reflection, serializing the struct into JSON in a shape identical to the golang structs.  This means it encounters a "Validators" field (with no json name specified) so this gets serialized with exactly the same name as the golang struct field, "Validators".

By contrast, even though the pulsar type has the same shape, it is traversed differently by protoreflect.  The "Validators" field is essentially ignored and flattened into oneof field.  So I needed to fabricate this encapsulating field, "Validators" somehow for protoreflect.  oneof_field_name is communicating the name of that field.

Another option is to defer to convention, but this would mean string munging on a potentially_compound_word to find equivalence with the golang struct field name.

Reply:
> Another option is to defer to convention, but this would mean string munging on a potentially_compound_word to find equivalence with the golang struct field name.

I'm also in favor of convention here. From what I can tell, the golang struct field name should always be PascalCase(oneof_field). Do you see a risk in using this convention?

Reply:
>Do you see a risk in using this convention?
No.  We can proceed with convention based here too since it will make migration easier.